### ISSUE TITLE
[C++] Pure ScalarFunctions called with no arguments should return scalar

### ISSUE THREAD

**Issue by bkietz** [apache/arrow#40510]
### Describe the enhancement requested

When called with no arguments, the desired output length of scalar functions cannot be determined in general. (For some functions such as `random` we can exploit default preallocation of fixed width output to inform the kernel how much output to generate, but this approach can't work for all functions.) However for the majority of scalar functions which are pure (their output is completely determined by their arguments), there is exactly one value corresponding to empty arguments. In the context of expression evaluation, returning this value as a scalar indicates that the value should be broadcast to whatever length is required for the enclosing expression. I think this is a more reasonable return value than an empty null array, which is what is returned by for example `pa.max_element_wise()`.

Note: explicit Function::is_impure flag added in https://github.com/apache/arrow/pull/40396

### Component(s)

C++

**Comment by felipecrv** [comment]
Note that I opened a PR renaming that to `is_pure()` https://github.com/apache/arrow/pull/40608

**Comment by ZhangHuiGui** [comment]
> I think this is a more reasonable return value than an empty null array, which is what is returned by for example `pa.max_element_wise()`.

Agree, we should return scalar for pure functions with no arguments.

**Comment by ZhangHuiGui** [comment]
@bkietz @felipecrv  We can have more discussion in the PR?

**Comment by thisisnic** [comment]
This issue has been closed because there has been no activity for the 14 days since the stale warning. Feel free to reopen if this has been closed in error.
